spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,582 @@
+    Schema_IndexSfixed32,
+    Schema_GetSfixed32Count,
+    Schema_AddSfixed32,
+    Schema_AddSfixed32List,
+    Schema_GetSfixed32List,
+);
+impl_primitive_field!(
+    i64,
+    SchemaSfixed64,
+    Schema_GetSfixed64,
+    Schema_IndexSfixed64,
+    Schema_GetSfixed64Count,
+    Schema_AddSfixed64,
+    Schema_AddSfixed64List,
+    Schema_GetSfixed64List,
+);
+impl_primitive_field!(
+    u32,
+    SchemaEnum,
+    Schema_GetEnum,
+    Schema_IndexEnum,
+    Schema_GetEnumCount,
+    Schema_AddEnum,
+    Schema_AddEnumList,
+    Schema_GetEnumList,
+);
+
+impl SchemaType for EntityId {
+    type RustType = Self;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let id = unsafe { Schema_GetEntityId(object.internal, field) };
+        Self { id }
+    }
+}
+
+impl SchemaIndexType for EntityId {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetEntityIdCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let id = unsafe { Schema_IndexEntityId(object.internal, field, index) };
+        Self { id }
+    }
+}
+
+impl SchemaListType for EntityId {
+    fn get_field_list(object: &SchemaObject, field: FieldId, data: &mut Vec<Self::RustType>) {
+        let count = Self::field_count(object, field) as usize;
+
+        // Ensure that there is enough capacity for the elements in the schema field.
+        if data.capacity() < count {
+            data.reserve(count - data.capacity());
+        }
+
+        // Replace the contents of `data` with the list of values in the schema field.
+        unsafe {
+            data.set_len(count);
+            Schema_GetEntityIdList(object.internal, field, data.as_mut_ptr() as *mut _);
+        }
+    }
+}
+
+impl SchemaType for bool {
+    type RustType = Self;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let raw = unsafe { Schema_GetBool(object.internal, field) };
+        raw != 0
+    }
+}
+
+impl SchemaIndexType for bool {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetBoolCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let raw = unsafe { Schema_IndexBool(object.internal, field, index) };
+        raw != 0
+    }
+}
+
+// TODO: This is only valid if schema bools are guaranteed to be 0 or 1. If schema
+// bools are allowed to have other values, this could result in undefined behavior
+// and should be removed.
+impl<T: SchemaIndexType> SchemaType for Option<T> {
+    type RustType = Option<T::RustType>;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let count = T::field_count(object, field);
+        match count {
+            0 => None,
+            1 => Some(T::from_field(object, field)),
+            _ => panic!(
```

Comment:
Ah, that's interesting! I had assumed that the runtime would reject malformed component data, but I guess not? If that's not the case, at least, then I agree that we shouldn't panic in this case. This would be a good thing to clarify in the serialization docs, though. I'll add a question about it to the forum thread as a reminder 😁